Source/PeleLM.cpp

```diff
@@ -6148,98 +6146,114 @@ PeleLM::compute_scalar_advection_fluxes_and_divergence (const MultiFab& Force,
   //
   // HERE IS THE EB PROCEDURE
   //
   //////////////////////////////////////
 
-
   // Initialize accumulation for rho = Sum(rho.Y)
-  for (int d=0; d<BL_SPACEDIM; d++) {
+  for (int d=0; d<AMREX_SPACEDIM; d++) {
     EdgeState[d]->setVal(0);
     EdgeFlux[d]->setVal(0);
   }
 
   // Advect RhoY  
   {
-    Vector<BCRec> math_bcs(nspecies);
-    math_bcs = fetchBCArray(State_Type, first_spec,nspecies);
-
-     MOL::ComputeAofs( *aofs, first_spec, nspecies, Smf, rhoYcomp,
+     Vector<BCRec> math_bcs(NUM_SPECIES);
+     math_bcs = fetchBCArray(State_Type, first_spec,NUM_SPECIES);
+     MOL::ComputeAofs( *aofs, first_spec, NUM_SPECIES, Smf, rhoYcomp,
                        D_DECL(u_mac[0],u_mac[1],u_mac[2]),
                        D_DECL(*EdgeState[0],*EdgeState[1],*EdgeState[2]), first_spec, false,
-                       D_DECL(*EdgeFlux[0],*EdgeFlux[1],*EdgeFlux[2]), first_spec,                    
+                       D_DECL(*EdgeFlux[0],*EdgeFlux[1],*EdgeFlux[2]), first_spec,
                        math_bcs, geom );
-
-
-    EB_set_covered(*aofs, 0.);
-    
+     EB_set_covered(*aofs, 0.);
   }
-  
-  // Set flux, flux divergence, and face values for rho as sums of the corresponding RhoY quantities
+
+   // Set flux, flux divergence, and face values for rho as sums of the corresponding RhoY quantities
 #ifdef _OPENMP
-#pragma omp parallel
+#pragma omp parallel if (Gpu::notInLaunchRegion())
 #endif
-  {
-    for (MFIter S_mfi(Smf,true); S_mfi.isValid(); ++S_mfi)
-    {
+   for (MFIter S_mfi(Smf,TilingIfNotGPU()); S_mfi.isValid(); ++S_mfi)
+   {
       const Box bx = S_mfi.tilebox();
+
       // this is to check efficiently if this tile contains any eb stuff
       const EBFArrayBox& in_fab = static_cast<EBFArrayBox const&>(Smf[S_mfi]);
       const EBCellFlagFab& flags = in_fab.getEBCellFlagFab();
 
       if(flags.getType(amrex::grow(bx, 0)) == FabType::covered)
       {
-        (*aofs)[S_mfi].setVal<RunOn::Host>(0,bx,Density,1);
-        for (int d=0; d<AMREX_SPACEDIM; ++d)
-        {
-          const Box& ebox = S_mfi.nodaltilebox(d);
-
-          (*EdgeState[d])[S_mfi].setVal<RunOn::Host>(0.,ebox,0,NUM_STATE);
-          (*EdgeFlux[d])[S_mfi].setVal<RunOn::Host>(0.,ebox,0,NUM_STATE);
-        }
-     }
+         auto const& adv_rho   = aofs->array(S_mfi,Density);  
+         amrex::ParallelFor(bx, [ adv_rho ]
+         AMREX_GPU_DEVICE (int i, int j, int k) noexcept
+         {
+            adv_rho(i,j,k) = 0.0; 
+         });
+         for (int dir=0; dir<AMREX_SPACEDIM; ++dir)
+         {
+            const Box& ebx = S_mfi.nodaltilebox(dir);
+            auto const& state_ed = EdgeState[dir]->array(S_mfi,0);
+            auto const& fluxes   = EdgeFlux[dir]->array(S_mfi,0);
+            amrex::ParallelFor(ebx, NUM_STATE, [state_ed,fluxes]
+            AMREX_GPU_DEVICE (int i, int j, int k, int n) noexcept
+            {
+               state_ed(i,j,k,n) = 0.0; 
+               fluxes(i,j,k,n) = 0.0; 
```

Comment:
Isn't setVal fully safe and GPU-ready?  is this just because there are two operations together so you are making one kernel launch?

Reply:
It's partly because there are two operations and also because I don't really know if there is a setVal<RunOn::Host_Device> ?